### PR TITLE
Use extensible nodes

### PR DESCRIPTION
**Pull request by jasonmp85**
Heavily based on two old commits by @anarazel, this introduces macros to wrap differences between 9.5 and 9.6 read- and out- functions for our custom nodes, allowing better code reuse for those functions.

Additionally, if running on PostgreSQL 9.6, this change will register extensible node types for all our custom nodes, permitting us to rely on the standard PostgreSQL `nodeToString`, `stringToNode`, and `readNode` functions.

Long-term, this change lets us avoid copy-pasting a ton of `readfuncs` code as well as well as some other `static` methods we've been copying in the past.

### CODE REVIEW

File: src/backend/distributed/utils/citus_nodefuncs.c [hunk 1]
```diff
@@ -307,3 +329,53 @@ citus_extradata_container(PG_FUNCTION_ARGS)
...
+	DEFINE_NODE_METHODS(MultiSelect),
+	DEFINE_NODE_METHODS(MultiTable),
+	DEFINE_NODE_METHODS(MultiJoin),
+	DEFINE_NODE_METHODS(MultiPartition),
+	DEFINE_NODE_METHODS(MultiCartesianProduct),
+	DEFINE_NODE_METHODS(MultiExtendedOp)
+};
+#endif
+
+void
```
Comment (anarazel) [inline src/backend/distributed/utils/citus_nodefuncs.c]: Hm, can we add a StaticAssert verifying that ExtensibleNodeMethods has the same length as CitusNodeTagNamesD?

File: src/include/distributed/citus_nodes.h [hunk 2]
```diff
@@ -17,9 +17,10 @@
  *
```
Comment (anarazel) [inline src/include/distributed/citus_nodes.h]: Can we add a two sentence overview about what has to be done to add a new node type here?

File: src/backend/distributed/utils/citus_outfuncs.c [hunk 3]
```diff
@@ -39,9 +39,13 @@
...
 
 /* Write the label for the node type */
+#if (PG_VERSION_NUM >= 90600)
+#define WRITE_NODE_TYPE(nodelabel) \
+	(void) 0
+#else
 #define WRITE_NODE_TYPE(nodelabel) \
 	appendStringInfoString(str, nodelabel)
-
+#endif
```
Comment (jasonmp85) [inline src/backend/distributed/utils/citus_outfuncs.c]: Added.

File: src/backend/distributed/utils/citus_nodefuncs.c (hunk 1 above)
Comment (jasonmp85) [inline src/backend/distributed/utils/citus_nodefuncs.c]: Sure.

File: src/include/distributed/citus_nodes.h [hunk 4]
```diff
@@ -69,5 +121,6 @@ typedef enum CitusNodeTag
  */
 #define CitusMakeNode(_type_) ((_type_ *) CitusNewNode(sizeof(_type_),T_##_type_))
```
Comment (jasonmp85) [inline src/include/distributed/citus_nodes.h]: Sure; also moved `CitusIsA`

File: src/include/distributed/citus_nodes.h (hunk 2 above)
Comment (jasonmp85) [inline src/include/distributed/citus_nodes.h]: Done, though with more than two sentences.

**Comment by jasonmp85** [comment]
Addressed all feedback, @anarazel, though the tests won't pass until I get a 9.6-compatible HLL package into packagecloud.

**Comment by jasonmp85** [comment]
@anarazel — tests all passing now, BTW

File: src/include/distributed/citus_nodes.h [hunk 5]
```diff
@@ -3,6 +3,26 @@
...
+ *     converting the node to a string
+ *
+ *   * Implement a 'readfunc' for the node in citus_readfuncs.c, using
+ *     the macros defined within that file. This function will handle
+ *     converting strings into instances of the node
+ *
+ *   * Use DEFINE_NODE_METHODS within the nodeMethods array (near the
+ *     bottom of citus_nodefuncs.c) to register the node in PostgreSQL
+ *
  * Copyright (c) 2012-2016, Citus Data, Inc.
```
Comment (anarazel) [inline src/include/distributed/citus_nodes.h]: Nice.

File: src/backend/distributed/utils/citus_outfuncs.c [hunk 6]
```diff
@@ -81,9 +87,15 @@
...
 /* Write a character-string (possibly NULL) field */
+#if (PG_VERSION_NUM >= 90600)
+#define WRITE_STRING_FIELD(fldname) \
+	(appendStringInfo(str, " :" CppAsString(fldname) " "), \
+	 outToken(str, node->fldname))
+#else
 #define WRITE_STRING_FIELD(fldname) \
 	(appendStringInfo(str, " :" CppAsString(fldname) " "), \
 	 _outToken(str, node->fldname))
+#endif
```
Comment (anarazel) [inline src/backend/distributed/utils/citus_outfuncs.c]: Hm, can't we get rid of that distinction? Why don't we just have our copy of _outToken as outToken() for 9.5?

File: src/backend/distributed/utils/citus_nodefuncs.c [hunk 7]
```diff
@@ -307,3 +329,57 @@ citus_extradata_container(PG_FUNCTION_ARGS)
...
+
+#if (PG_VERSION_NUM >= 90600)
+
+/* *INDENT-OFF* */
+#define DEFINE_NODE_METHODS(type) \
+	{ \
+		#type, \
+		sizeof(type), \
+		NULL, \
+		NULL, \
```
Comment (anarazel) [inline src/backend/distributed/utils/citus_nodefuncs.c]: Hm, these are actually mandatory, so until we need them, we should probably use the same function in all of them that just errors out.

File: src/backend/distributed/utils/citus_outfuncs.c (hunk 6 above)
Comment (jasonmp85) [inline src/backend/distributed/utils/citus_outfuncs.c]: Done.

**Comment by jasonmp85** [comment]
@anarazel Updated once again.

File: src/backend/distributed/utils/citus_outfuncs.c [hunk 8]
```diff
@@ -217,39 +223,52 @@ _outDatum(StringInfo str, Datum value, int typlen, bool typbyval)
 		}
 	}
 }
-
```
Comment (anarazel) [inline src/backend/distributed/utils/citus_outfuncs.c]: removes existing newline. Change on your own discretion.

File: src/backend/distributed/utils/citus_outfuncs.c [hunk 9]
```diff
@@ -217,39 +223,52 @@ _outDatum(StringInfo str, Datum value, int typlen, bool typbyval)
...
 static void
-_outMultiBinaryNode(StringInfo str, const MultiBinaryNode *node)
+OutMultiBinaryNodeFields(StringInfo str, const MultiBinaryNode *node)
 {
 	WRITE_NODE_FIELD(leftChildNode);
 	WRITE_NODE_FIELD(rightChildNode);
 }
 
+#define WRITE_LOCALS(nodeTypeName) \
+		const nodeTypeName *node = (const nodeTypeName *) raw_node
```
Comment (anarazel) [inline src/backend/distributed/utils/citus_outfuncs.c]: Why's this in the middle of the file?

Comment (jasonmp85) [inline src/backend/distributed/utils/citus_outfuncs.c]: No clue! I didn't write that commit.

**Pull request merged**